cluster/src/main/java/com/hartwig/pipeline/calling/structural/StructuralCaller.java

```diff
@@ -88,69 +95,46 @@ public StructuralCallerOutput run(final SomaticRunMetadata metadata, final Align
-                tumorBam.getLocalTargetPath(),
-                referenceGenomePath,
-                jointName,
-                configurationFile,
-                blacklist);
-
-        IdentifyVariants calling = commandFactory.buildIdentifyVariants(referenceBam.getLocalTargetPath(),
-                tumorBam.getLocalTargetPath(),
-                assemblyResult.assemblyBam(),
-                referenceGenomePath, configurationFile, blacklist);
-
-        Annotation.AnnotationResult annotationResult =
-                new Annotation(commandFactory).initialise(referenceBam.getLocalTargetPath(),
-                        tumorBam.getLocalTargetPath(),
-                        assemblyResult.assemblyBam(),
-                        calling.resultantVcf(),
-                        referenceGenomePath,
-                        tumorSampleName,
-                        configurationFile,
-                        blacklist);
-
-        Filter.FilterResult filterResult = new Filter().initialise(annotationResult.annotatedVcf(), tumorSampleName);
-
-        bash.addCommands(preprocessedRefSample.commands())
-                .addCommands(preprocessedTumorSample.commands())
-                .addCommands(assemblyResult.commands())
-                .addCommand(calling)
-                .addCommands(annotationResult.commands())
-                .addCommands(filterResult.commands());
 
+        String refBamPath = referenceBam.getLocalTargetPath();
+        String tumorBamPath = tumorBam.getLocalTargetPath();
+        new Preprocess(refBamPath, referenceWorkingDir, referenceSampleName, referenceGenomePath, preprocessSvOutputReferenceBam)
+                .apply(SubStageInputOutput.of(referenceSampleName, OutputFile.empty(), bash));
+        new Preprocess(tumorBamPath, tumorWorkingDir, tumorSampleName, referenceGenomePath, preprocessSvOutputTumorBam)
+                .apply(SubStageInputOutput.of(tumorSampleName, OutputFile.empty(), bash));
+
+        Assemble assemble = new Assemble(refBamPath, tumorBamPath, jointName, referenceGenomePath, configurationFile, blacklist);
+        SubStageInputOutput result =
+                assemble.andThen(new Calling(refBamPath, tumorBamPath, "", referenceGenomePath, configurationFile, blacklist))
+                        .andThen(new Annotation(referenceBam.getLocalTargetPath(),
+                                tumorBam.getLocalTargetPath(),
+                                assemble.completedBam(),
+                                "",
+                                referenceGenomePath,
+                                jointName,
+                                configurationFile,
+                                blacklist))
+                        .apply(SubStageInputOutput.of(jointName, OutputFile.empty(), bash));
+        Filter.FilterResult filterResult = new Filter().initialise(result.outputFile().path(), tumorSampleName);
```

Comment:
Let's make this a function as well and pass in the final output file names, not perfect but then we can keep it all consistent (and think about returning multiple files later)